### PR TITLE
Add proper Stable Presentation Support

### PR DESCRIPTION
**Pull request by rochala**
This PR removes mtags3 code from metals in favour of wrapping Stable Presentation Compiler along with its nightly releases CI. At first, it will fail because of `coursier-interfaces` binary incompatible changes, but after https://github.com/lampepfl/dotty/pull/18264 is merged, the following nightly should be a successful run.

This PR also optimises the way we release `mtags-shared` module. It doesn't need cross version set to full, as `scala3-presentation-compiler` compiles it from 2.13 sources, and it does not depend on Scala library / compiler.

This PR also needs a better solution for handling Scala versions (3.3.0 should not be deprecated, as user will receive a warning).

### CODE REVIEW

**Comment by rochala** [comment]
The nightly CI will now enforce to backport changes to dotty, as the tests won't pass.
This PR is the example of this, it won't pass the CI until https://github.com/lampepfl/dotty/pull/18296 is merged and released as nightly. This will ensure that `scala3-presentation-compiler` is always up-to-date and will put backporting effort on maintainers.

File: build.sbt [hunk 1]
```diff
@@ -339,9 +352,19 @@ val mtagsSettings = List(
     val scalaVersionsWithSpecialCompat = Set("2.13.9", "2.13.10", "2.13.11")
     if (scalaVersionsWithSpecialCompat(scalaVersion.value))
       current.filter(f => f.getName() != "scala-2.13")
+    else if (isScala3PresentationCompiler(scalaVersion.value))
+      List(base / s"scala-3-presentation-compiler")
     else
       current
   },
+  Test / unmanagedSourceDirectories ++= {
```
Comment (tgodzik) [inline build.sbt:360]: What classes do we need this for?

File: build.sbt [hunk 2]
```diff
@@ -242,9 +251,8 @@ lazy val mtagsShared = project
   .settings(
     moduleName := "mtags-shared",
     crossTarget := target.value / s"scala-${scalaVersion.value}",
-    crossScalaVersions := {
-      V.supportedScalaVersions ++ V.nightlyScala3Versions
-    },
+    // Dotty depends on Scala 2.13 for compatibility guarantees for from-source compilation.
```
[the remark this answers is not in the document]
Comment (rochala) [reply build.sbt:254]: Well, I'm not sure about this either. For now we are compiling it from 2.13.11 sources, and remember that when we remove mtags3 source code in the future, we will stop releasing mtags-shard (nothing in this project will depend on them)

File: build.sbt (hunk 1 above)
Comment (rochala) [reply build.sbt:360]: We need this for testing. Cross tests rely on mtags, docstring, global indexes and jdksources.

File: build.sbt [hunk 3]
```diff
@@ -364,11 +387,35 @@ lazy val mtags3 = project
...
+  .enablePlugins(BuildInfoPlugin)
+
 lazy val mtags = project
   .settings(
     sharedSettings,
     mtagsSettings,
     moduleName := "mtags",
+    projectDependencies := {
+      // mtags-shared are included in scala3-presentation-compiler module as sources
+      if (isScala3PresentationCompiler(scalaVersion.value)) {
```
[the remark this answers is not in the document]
Comment (rochala) [reply build.sbt:415]: Nah, the mtags-shared in the compiler are part of `scala3-presentation-compiler` artifact as we compiled them from sources (there is no such concept in sbt, so we are compiling them as part of `scala3-presentation-compiler`). They are not present in dependency tree.

File: build.sbt [hunk 4]
```diff
@@ -648,7 +696,7 @@ lazy val mtest = project
       scalaVersion.value,
     ),
   )
-  .dependsOn(mtags)
+  .dependsOn(mtags % "compile->test")
```
[the remark this answers is not in the document]
Comment (rochala) [reply build.sbt:699]: Because we add `scala` only as test dependencies, we have to specify that `mtest` compilation depends on `Test` of mtags3WithPresentationCompiler.
```scala
  Test / unmanagedSourceDirectories ++= {
    val base = (Compile / sourceDirectory).value
    if (isScala3PresentationCompiler(scalaVersion.value)) {
      List(base / s"scala")
    } else {
      Nil
    }
  },
```

**Comment by rochala** [comment]
This PR requires backports in dotty. It won't pass the CI until the next day. We should ignore this CI for now, because any new test added before new scala3-presentation-compiler is released are going to fail.

File: build.sbt (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply build.sbt:254]: Ach ok, now I see the `filterNot`, didn't see it before.

File: build.sbt (hunk 1 above)
Comment (tgodzik) [reply build.sbt:362]: Could we not just add this directory to the test scope in the tests? This looks hard to follow since you add the directory here, but then it's actually only needed for tests.

File: build.sbt (hunk 4 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply build.sbt:701]: Maybe do it here:
```
Test/ unmanagedSourceDirectories ++= {
      mtags.base  / s"scala-3-wrapper"
    }
```
will that work?

**Comment by tgodzik** [comment]
> This PR requires backports in dotty. It won't pass the CI until the next day. We should ignore this CI for now, because any new test added before new scala3-presentation-compiler is released are going to fail.

Let's make sure that this gets merged with green PR though, we can stop merging anything problematic for a day. The actual workflow will need to be to merge things into the compiler first and then backport here :thinking:

**Review by tgodzik — APPROVED** [review]
LGTM!
